Adding in Wrap functions for FSDP from Fairscale

sisilmehta2000 opened the pull request:
Summary: as title

Test Plan: buck test mode/dev-nosan //caffe2/test/distributed/fsdp:wrap --keep-going

Differential Revision: D31936404



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma reviewed (commented):
Looks good to me overall to land this as a private API at first to unblock and then formalize the API. Although still have the question re: commented out test.

@zhaojuanmao can you also take a look?

zhaojuanmao commented on test/distributed/fsdp/test_wrap.py:
```diff
@@ -0,0 +1,158 @@
+# Owner(s): ["oncall: distributed"]
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
```
nit: remove this

zhaojuanmao reviewed (commented):
looks good, waiting for the tests to pass

Commit:
Adding in Wrap functions for FSDP from Fairscale (#67292)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/67292

as title

Test Plan: buck test mode/dev-nosan //caffe2/test/distributed/fsdp:wrap --keep-going

Differential Revision: D31936404

fbshipit-source-id: dfa1d779509f8c0438889fc626ddd94b2b745ec0

rohan-varma reviewed (approved):
LGTM, please wait for the ROCM CI to clear before landing, thanks for this!
